### PR TITLE
Update to Azure OpenAI version 2024-02-01

### LINKED ISSUE #197: Get requests to `/conversation` endpoints should return 405 Method Not Allowed
> Please provide us with the following information:
> ---------------------------------------------------------------

The `/api/conversation/azure_byod` and `/api/conversation/custom` endpoints supposedly allow GET requests, however they both respond with an error, as they are expecting request bodies. They should instead return a 405 Method Not Allowed.

### This issue is for a: (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Any log messages given by the failure
```
{
  "error": "Exception in /api/conversation/azure_byod. See log for more details."
}
```

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

### LINKED ISSUE #413: Use custom subdomain for form recognizer and content safety
# Describe the bug
At present, the form recognizer and content safety endpoints use the regional endpoints, e,g. `https://uksouth.api.cognitive.microsoft.com/`. These endpoints support API key authentication but not RBAC. 

We should switch to the custom subdomain to use RBAC with these services. The custom subdomain also supports API key authentication.

# Expected behavior
Documents uploaded should be processed successfully

---

# Debugging information

## Steps to reproduce
1. Navigate to Admin site
2. Go to Ingest Data tab
3. Upload document (e.g. PDF)
4. View `batch_push_results` Azure functions logs

## Logs

```
2024-03-12T09:53:45Z   [Error]   Error adding embeddings for ******************: Error: Traceback (most recent call last):
  File "/home/site/wwwroot/utilities/helpers/AzureFormRecognizerHelper.py", line 78, in begin_analyze_document_from_url
    poller = self.document_analysis_client.begin_analyze_document_from_url(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/ai/formrecognizer/_document_analysis_client.py", line 198, in begin_analyze_document_from_url
    return _client_op_path.begin_analyze_document(  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/ai/formrecognizer/_generated/v2023_07_31/operations/_document_models_operations.py", line 518, in begin_analyze_document
    raw_result = self._analyze_document_initial(  # type: ignore
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/ai/formrecognizer/_generated/v2023_07_31/operations/_document_models_operations.py", line 443, in _analyze_document_initial
    raise HttpResponseError(response=response)
azure.core.exceptions.HttpResponseError: (BadRequest) Please provide a custom subdomain for token authentication, otherwise API key is required. (Please see https://aka.ms/cogsvc-authenticatewithtoken)
Code: BadRequest
Message: Please provide a custom subdomain for token authentication, otherwise API key is required. (Please see https://aka.ms/cogsvc-authenticatewithtoken)
. Error: (BadRequest) Please provide a custom subdomain for token authentication, otherwise API key is required. (Please see https://aka.ms/cogsvc-authenticatewithtoken)
Code: BadRequest
Message: Please provide a custom subdomain for token authentication, otherwise API key is required. (Please see https://aka.ms/cogsvc-authenticatewithtoken)
```

---

# Tasks

_To be filled in by the engineer picking up the issue_

### LINKED ISSUE #552: Update to latest stable Azure OpenAI API version

### LINKED ISSUE #554: Azure function app unable to generate embeddings for documents when using RBAC
# Describe the bug
Azure function app is unable to generate embeddings for documents when using RBAC. It lacks the permissions to both generate an SAS to download the file, and to add the metadata to the document.

# Expected behavior
After uploading a document with the Admin app, embeddings should be generated and stored in Azure AI Search. The metadata on the document in the Storage Account should also be updated

---

# Debugging information

## Steps to reproduce

* Deploy application with `AZURE_AUTH_TYPE=rbac`
* Upload document through Admin app
* Observe function app logs

## Logs

```
Error adding embeddings for https://************.blob.core.windows.net/documents/PressReleaseFY23Q4.docx?**********************: File is not a zip file
```

---

# Tasks

_To be filled in by the engineer picking up the issue_

### PR DESCRIPTION
**Pull request by cecheta**
Closes #554, closes #197, closes #552, closes #413 

## Purpose

* This PR updates to the latest stable version of Azure OpenAI API version, `2024-02-01`.
  * There are no changes required to the `/api/conversation/custom` endpoint, but all the changes are to the `/api/conversation/azure_byod` endpoint. This endpoint is not called by the frontend at all (and was not currently working anyway)
* Update unit tests to more closely match streamed response
* Fix issues regarding RBAC and unimplemented GET methods

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Update code [here](https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator/blob/main/code/frontend/src/pages/chat/Chat.tsx#L82) to use `conversationApi()` function
* Experiment with RBAC/non-RBAC, streaming/non-streaming, use data/no data

## What to Check
Verify that the following are valid
* Response back from the API are displayed correctly in the frontend

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -35,7 +35,10 @@
 			"settings": {
 				"azureFunctions.projectSubpath": "code/backend/batch",
                 "python.defaultInterpreterPath": "/usr/local/bin/python",
-				"python.pythonPath": "/usr/local/bin/python"
+				"python.pythonPath": "/usr/local/bin/python",
+				"files.insertFinalNewline": true,
+				"files.trimFinalNewlines": true,
```
Comment (ross-p-smith) [inline .devcontainer/devcontainer.json:40]: 💖

File: infra/app/function.bicep [hunk 2]
```diff
@@ -112,6 +112,16 @@ module searchRoleFunction '../core/security/role.bicep' = if (authType == 'rbac'
   }
 }
 
+// Storage Blob Data Contributor
+module storageRoleFunction '../core/security/role.bicep' = if (authType == 'rbac') {
+  name: 'storage-role-function'
+  params: {
+    principalId: function.outputs.identityPrincipalId
```
Comment (ross-p-smith) [inline infra/app/function.bicep:119]: Is this the bug @tanya-borisova is working on?

**Review by ross-p-smith — APPROVED** [review]
Just Wow! Loads in here and really great work!

File: code/app.py [hunk 3]
```diff
@@ -128,37 +121,54 @@ def stream_with_data(body, headers, endpoint):
...
+                    {
+                        "content": "",
+                        "end_turn": False,
+                        "role": "tool",
+                    },
+                    {
+                        "content": "",
+                        "end_turn": False,
+                        "role": "assistant",
+                    },
```
Comment (superhindupur) [inline code/app.py:136]: Could you explain this change?

**Review by superhindupur — APPROVED** [review]

File: code/app.py (hunk 3 above)
Comment (cecheta) [reply code/app.py:136]: Discussed over a call

**Comment by andrewwiebe** [comment]
I havent looked over these changes in detail but we are actually using the azure_byod route. We switched away from the custom to the azure_byod route as it was simpler to make changes to the openAI call and passing data source details and a filter string.

Should we assume this route will no longer functional or be updated when moving to the new api versions?

**Comment by ross-p-smith** [comment]
Hey @andrewwiebe - both routes should be functional. We needed to update the OpenApi versions but this change only effected the byod route. If you think something is broken with the byod route then please raise a bug and we will prioritise. In summary - both routes are supported.

**Pull request merged**